### PR TITLE
errors: `Wrap()` for unformatted, `Wrapf()` for formatted [BREAKING]

### PR DESCRIPTION
**Pull request by amery**
and choose `golangci-lint` and `revive` version based on the Go version

## Summary by CodeRabbit

- **New Features**
	- Introduced a new error wrapping function `Wrapf` for enhanced flexibility in error annotations.
	- Added a script for dynamic version retrieval of Go tools.

- **Improvements**
	- Expanded testing coverage by including an additional Go version in the workflow.
	- Updated tool versions in the build process for better compatibility.
	- Simplified error handling processes in various functions, enhancing clarity and usability.

- **Documentation**
	- Restructured the "Errors" section in the README to provide clearer guidance on error handling mechanisms.
	- Added a new subsection on "Wrappers" detailing error wrapping functions and their usage.

### CODE REVIEW

File: internal/build/get_version.sh [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+	exit 0
+fi
+
+GO_MAJOR=$(echo "$GO_VERSION" | cut -d. -f1)
+GO_MINOR=$(echo "$GO_VERSION" | cut -d. -f2)
+GO_VER=$(( GO_MAJOR * 1000 + GO_MINOR ))
+
+X=$(echo "$1" | cut -d. -f1)
+Y=$(echo "$1" | cut -d. -f2)
+VER=$(( X * 1000 + Y))
```
[the remark this answers is not in the document]
Comment (karasz) [reply internal/build/get_version.sh:17]: Probably this is effective.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -12,7 +12,7 @@ jobs:
     if: (github.event_name == 'push' || github.event.pull_request.head.repo.full_name != github.repository)
     strategy:
       matrix:
-        go: [ '1.20', '1.21', '1.22' ]
+        go: [ '1.20', '1.21', '1.22', '1.23' ]
```
Comment (karasz) [inline .github/workflows/build.yml:15]: `1.20` is not needed in fact. We are ok if we maintain 2 versions, 3 is very nice.

File: internal/build/get_version.sh [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+if [ $GO_VER -ge $VER -a $# -gt 0 ]; then
+	for ver; do
+		[ $GO_VER -gt $VER ] || break
+
+		VER=$(( VER + 1 ))
+	done
+
+	echo "$ver"
+	exit 0
+fi
```
[the remark this answers is not in the document]
Comment (karasz) [reply internal/build/get_version.sh:30]: The shellcheck is kinda ok. I do not care for the rest of the proposed changes.

**Review by karasz — CHANGES_REQUESTED** [review]

**Comment by amery** [comment]
@karasz we test 1.20 and 1.21 not because it's nice but because go.mod says they are supported. dropping 1.20 would be a different change that can also be part of the 0.15 bump, but afaik it's not required by anything

File: internal/build/get_version.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (amery) [reply internal/build/get_version.sh:30]: > @karasz, understood. Here's the change to address the Shellcheck warning:
> 
> ```shell
> if [ $GO_VER -ge $VER ] && [ $# -gt 0 ]; then
> ```
> 
> This replaces `-a` with `&&` as recommended by Shellcheck.
> 
> (_/) ( •_•) /☔
> 
> ✏️ Learnings added

@coderabbitai why do you want me to replace the `-a` with two `[` calls?

File: .github/workflows/build.yml (hunk 2 above)
Comment (amery) [reply .github/workflows/build.yml:15]: `go.mod` says `1.20` so `1.20` needs testing. nothing on this commit requires dropping `1.20`. if we want `1.20` dropped for `v0.15` I believe that should be a different PR

**Comment by amery** [comment]
I'll split the PR

**Comment by amery** [comment]
`get_version()` change moved to #75

**Comment by amery** [comment]
switching this to draft, to address the spin-off PRs first.

**Comment by amery** [comment]
here we go again. v0.15 making golangci-lint v1.60+ happy about client dual use of `Wrap()`

@karasz review
@coderabbitai review
@deepsource-io review

File: README.md [hunk 4]
```diff
@@ -61,11 +61,14 @@ should be on a subdirectory, it shouldn't be here.
...
-* `Wrap(err, note)` with a simple string, or with a formatted note `Wrap(err, format, args...)`.
-* And the same for `QuietWrap(err, note)` or `QuietWrap(err, format, args...)`.
+There are three factories for `Unwrappable`, the standard `"note: error description"`,
+one for formatted notes, and a quiet one, not including the text of the original error
+unless unwrapped first.
+
+* `Wrap(err, note)` with a simple string,
+* `Wrapf(err, format, args...)` when using a formatted note,
+* and `QuietWrapf(err, format, args...)` for a formatted errors not including
+  the wrapped message in the text.
```
[the remark this answers is not in the document]
Comment (amery) [reply README.md]: fixed

File: compounderror.go [hunk 5]
```diff
@@ -102,7 +102,7 @@ func (w *CompoundError) Append(err error, note string, args ...any) {
 		err = errors.New(note)
 	case note != "":
 		// wrap
-		err = Wrap(err, "%s", note)
+		err = Wrap(err, note)
```
[the remark this answers is not in the document]
Comment (karasz) [reply compounderror.go:105]: @coderabitai `Wrap` function is defined in the `errors.go` file in the same package.

**Review by karasz — APPROVED** [review]

**Comment by amery** [comment]
`v0.15.0` tagged

**Pull request merged**